File: PetriNetPlugin/src/org/workcraft/plugins/petri/tools/SimulationTraceTable.java
package org.workcraft.plugins.petri.tools;

import java.awt.Component;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.List;

import javax.swing.JButton;
import javax.swing.JTable;
import javax.swing.ListSelectionModel;
import javax.swing.table.AbstractTableModel;
import javax.swing.table.TableCellRenderer;

import org.workcraft.Trace;
import org.workcraft.util.Function2;
import org.workcraft.util.Nothing;
import org.workcraft.util.Pair;

import pcollections.TreePVector;

public class SimulationTraceTable<State> {

	private TracePair traces = TracePair.createEmpty();

	Pair<Trace, Integer> getCurrentTrace() {
		if(traces.branchTrace.isEmpty())
			return Pair.of(traces.trace, traces.traceStep);
		else
			return Pair.of(traces.branchTrace, traces.branchStep);
	}
	
	public SimulationControl<SimulationState<State>> getSimControl() {
		return new SimulationControl<SimulationState<State>>() {
	
			@Override
			public boolean canFire(Nothing event) {
				Pair<Trace, Integer> curTrace = getCurrentTrace();
				Trace trace = curTrace.getFirst();
				int step = curTrace.getSecond();
				
				return step < trace.size() && model.canFire(trace.get(step));
			}
	
			@Override	
			public void fire(Nothing event) {
				quietStep();
			}
	
			@Override
			public boolean canUnfire(Nothing event) {
				Pair<Trace, Integer> curTrace = getCurrentTrace();
				Trace trace = curTrace.getFirst();
				int step = curTrace.getSecond();

				return  step > 0 && model.canUnfire(trace.get(step-1));
			}

			@Override
			public void unfire(Nothing event) {
				quietStepBack();
			}

			
			@Override
			public SimulationState<State> saveState() {
				return new SimulationState<State>(model.saveState(), traces.clone());
			}

			@Override
			public void loadState(SimulationState<State> state) {
				model.loadState(state.modelState);
				traces = state.traces.clone();
			}
		};
	};

	
	public SimulationTraceTable(SimulationModel<String, State> model, Function2<String, Boolean, Component> cellRenderer) {
		this.model = model;
		traceTable = create();
		
		loadFromClipboardButton = new JButton("from Clipb");
		saveToClipboardButton = new JButton("to Clipb");
		

		saveToClipboardButton.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				TracesClipboardOwner.saveToClipboard(traces);
			}

		});

		loadFromClipboardButton.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				traces = TracesClipboardOwner.loadFromClipboard();
			}
		});
		
		this.cellRenderer = cellRenderer;
	}
	
	
	final SimulationModel<String, State> model;
	
	private final JTable traceTable;
	private JButton saveToClipboardButton, loadFromClipboardButton;

	public List<Component> components() {
		return TreePVector.<Component>empty()
			.plus(traceTable)
			.plus(saveToClipboardButton)
			.plus(loadFromClipboardButton);
	}

	private boolean quietStepBack() {
		if (traces.branchStep > 0) {
			String event = traces.branchTrace.get(traces.branchStep - 1);

			if(!model.canUnfire(event))
				return false;
			
			model.unfire(event);
			
			traces.branchStep--;

			if (traces.branchStep == 0 && !traces.trace.isEmpty())
				traces.branchTrace.clear();
			return true;
		}

		if (traces.traceStep == 0)
			return false;

		String event = traces.trace.get(traces.traceStep - 1);
		
		if(!model.canUnfire(event))
			return false;
		model.unfire(event);
		
		traces.traceStep--;
		
		return true;
	}

	private boolean quietStep() {
		if(!traces.branchTrace.isEmpty()) { // If we have a branch-trace, use that. Otherwise, use the main trace.
			if (traces.branchStep < traces.branchTrace.size()) {
				String event = traces.branchTrace.get(traces.branchStep);
				
				if(!model.canFire(event))
					return false;
				model.fire(event);
				traces.branchStep++;

				return true;
			}
			else
				return false;
		}
		else {
			if (traces.traceStep == traces.trace.size())
				return false;
	
			String event = traces.trace.get(traces.branchStep);
			
			if(!model.canFire(event))
				return false;
			model.fire(event);
	
			traces.traceStep++;
			return true;
		}
	}
	
	public void userRequestedFire(String event) {
		String nextEvent = getNextEvent();
		
		if (!event.equals(nextEvent)) {
			while (traces.branchStep < traces.branchTrace.size())
				traces.branchTrace.remove(traces.branchStep);
	
			traces.branchTrace.add(event);
		}
		
		quietStep();
	}

	final Function2<String, Boolean, Component> cellRenderer;
	
	private JTable create() {
		final JTable traceTable = new JTable(new AbstractTableModel() {
			private static final long serialVersionUID = 1L;

			@Override
			public int getColumnCount() {
				return 2;
			}

			@Override
			public String getColumnName(int column) {
				if (column == 0)
					return "Trace";
				return "Branch";
			}

			@Override
			public int getRowCount() {
				int tnum = traces.trace.size();
				int bnum = traces.branchTrace.size();

				return Math.max(tnum, bnum + traces.traceStep);
			}

			@Override
			public Object getValueAt(int row, int col) {
				if (col == 0) {
					if (!traces.trace.isEmpty() && row < traces.trace.size())
						return traces.trace.get(row);
				} else {
					if (!traces.branchTrace.isEmpty() && row >= traces.traceStep && row < traces.traceStep + traces.branchTrace.size()) {
						return traces.branchTrace.get(row - traces.traceStep);
					}
				}
				return "";
			}
		});
		
		traceTable.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);

		traceTable.addMouseListener(new MouseListener() {

			@Override
			public void mouseClicked(MouseEvent e) {
				int column = traceTable.getSelectedColumn();
				int row = traceTable.getSelectedRow();

				if (column == 0) {
					if (traces.trace != null && row < traces.trace.size()) {

						boolean work = true;

						while (traces.branchStep > 0 && work)
							work = quietStepBack();
						while (traces.traceStep > row && work)
							work = quietStepBack();
						while (traces.traceStep < row && work)
							work = quietStep();
					}
				} else {
					if (traces.branchTrace != null && row >= traces.traceStep && row < traces.traceStep + traces.branchTrace.size()) {

						boolean work = true;
						while (traces.traceStep + traces.branchStep > row && work)
							work = quietStepBack();
						while (traces.traceStep + traces.branchStep < row && work)
							work = quietStep();
					}
				}
			}

			@Override
			public void mouseEntered(MouseEvent arg0) {

			}

			@Override
			public void mouseExited(MouseEvent arg0) {
			}

			@Override
			public void mousePressed(MouseEvent arg0) {
			}

			@Override
			public void mouseReleased(MouseEvent arg0) {
			}

		});

		traceTable.setDefaultRenderer(Object.class, new TableCellRenderer() {
			boolean isActive(int row, int column) {
				if (column == 0) {
					if (traces.branchTrace.isEmpty())
						return row == traces.traceStep;
				} else {
					if (row >= traces.traceStep && row < traces.traceStep + traces.branchTrace.size()) {
						return (row - traces.traceStep) == traces.branchStep;
					}
				}

				return false;
			}

			@Override
			public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
				if (!(value instanceof String))
					return null;
				return cellRenderer.apply((String) value, isActive(row, column));
			}
		});
		return traceTable;
	}

	public void setTraces(TracePair traces) {
		this.traces = traces;
	}


	public String getNextEvent() {
		Pair<Trace, Integer> currentTrace = getCurrentTrace();
		int i = currentTrace.getSecond();
		Trace t = currentTrace.getFirst();
		if(t.size() == i)
			return null;
		else
			return t.get(i);
	}
	
	public SimControl<String> asSimControl() {
		return new SimControl<String>() {
			
			@Override
			public void unfire() {
				SimulationTraceTable.this.quietStepBack();
			}
			
			@Override
			public String getNextEvent() {
				return SimulationTraceTable.this.getNextEvent();
			}
			
			@Override
			public void fire(String event) {
				userRequestedFire(event);
			}
			
			@Override
			public boolean canFire(String event) {
				return model.canFire(event);
			}
		};
	}
}

File: ScalaGraphEditorUtil/src/main/scala/org/workcraft/plugins/petri/tools/SimulationTraceTable.scala
package org.workcraft.plugins.petri.tools

import java.awt.Component
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.MouseEvent
import java.awt.event.MouseListener

import javax.swing.JButton
import javax.swing.JTable
import javax.swing.ListSelectionModel
import javax.swing.table.AbstractTableModel
import javax.swing.table.TableCellRenderer

import org.workcraft.swing.Swing
import Swing._

import SimulationControlPanel._

import scalaz._
import Scalaz._

class SimulationTraceTable[State]
  ( model : SimulationModel[Swing, String, State]
  , cellRenderer : (String, Boolean) => Component
  ) {
  val traceTable = create
  val loadFromClipboardButton = new JButton("from Clipb")
  val saveToClipboardButton = new JButton("to Clipb")
  saveToClipboardButton.addActionListener(new ActionListener {
      override def actionPerformed(e : ActionEvent) = TracesClipboardOwner.saveToClipboard(traces)
    })

  loadFromClipboardButton.addActionListener(new ActionListener {
    override def actionPerformed(e : ActionEvent) {
      traces = TracesClipboardOwner.loadFromClipboard
    }
  })

  var traces : TracePair = TracePair.createEmpty

  def getCurrentTrace : TraceStep = {
    if(traces.branch.trace.list.isEmpty)
      traces.main
    else
      traces.branch
  }

  def getSimControl : SimulationControl[Swing, SimulationState[State]] = new SimulationModel[Swing, Unit, SimulationState[State]]() {
    override def canFire(event : Unit) = unsafeToSwing {
      val TraceStep(trace, step) = getCurrentTrace
      step < trace.size && model.canFire(trace(step)).unsafePerformIO
    }
    override def fire(event : Unit) = unsafeToSwing(quietStep)
    override def canUnfire(event : Unit) = unsafeToSwing {
      val TraceStep(trace, step) = getCurrentTrace
      step > 0 && model.canUnfire(trace(step-1)).unsafePerformIO
    }
    override def unfire(event : Unit) = unsafeToSwing {
      quietStepBack
    }
    override def saveState : Swing[SimulationState[State]] = unsafeToSwing(new SimulationState[State](model.saveState.unsafePerformIO, traces))
    override def loadState(state : SimulationState[State]) = unsafeToSwing {
      model.loadState(state.modelState).unsafePerformIO
      traces = state.traces
    }
  }

  def components : List[Component] = List ( 
	  traceTable
          , saveToClipboardButton
          , loadFromClipboardButton
	  )

  def quietStepBack : Boolean = {
    def go(lens : Lens[TracePair, TraceStep]) : Boolean = {
      val TraceStep(trace, step) = lens.get(traces)
      val event = trace(step - 1)

      if(!model.canUnfire(event).unsafePerformIO)
	false
      else {
	model.unfire(event).unsafePerformIO
      
	traces = (TraceStep.step compose lens).mod(traces, _ - 1)
	true
      }
    }

    if (traces.branch.step > 0) {
      val res = go(TracePair.branch)
      
      if (res && traces.branch.step == 0 && !traces.main.trace.isEmpty)
	traces = (TraceStep.trace compose TracePair.branch).set(traces, Trace(Nil))

      res
    } else {
      if (traces.main.step == 0)
	false
      else 
	go(TracePair.main)
    }
  }

  def quietStep : Boolean = {
    if(!traces.branch.trace.isEmpty) { // If we have a branch-trace, use that. Otherwise, use the main trace.
      if (traces.branch.step < traces.branch.trace.size) {
	val event = traces.branch.trace(traces.branch.step)
    
	if(!model.canFire(event).unsafePerformIO) {println("can't fire " + event) ; 
	  return false}
	model.fire(event).unsafePerformIO

	traces = (TraceStep.step compose TracePair.branch).mod(traces, _ + 1)
	
	return true
      }
	else
	  return false
    }  else {
      if (traces.main.step == traces.main.trace.size)
	return false
	  
      val event = traces.main.trace(traces.branch.step)
	  
      if(!model.canFire(event).unsafePerformIO)
	return false
      model.fire(event).unsafePerformIO
	  
      traces = (TraceStep.step compose TracePair.main).mod(traces, _ + 1)
      return true
    }
  }
	
    def userRequestedFire(event : String) = {
      val nextEvent = getNextEvent
      println("firing " + event + " (next event : " + nextEvent + ")")
      if (!event.equals(nextEvent)) {
	traces = TracePair.branch.mod(traces, branch => {
	  TraceStep(Trace(branch.trace.list.take(branch.step) :+ event), branch.step)
	})
      }
      quietStep
    }

    def create : JTable = {
      val traceTable = new JTable(new AbstractTableModel {
	override def getColumnCount = 2

	override def getColumnName(column : Int) : String = {
	  if (column == 0)
	    return "Trace"
	  return "Branch"
	}
	override def getRowCount = {
	  val tnum = traces.main.trace.size
	  val bnum = traces.branch.trace.size

	  Math.max(tnum, bnum + traces.main.step)
	}

	override def getValueAt(row : Int, col : Int) : Object = {
	  if (col == 0) {
	    if (!traces.main.trace.isEmpty && row < traces.main.trace.size)
	      return traces.main.trace(row)
	  } else {
	    if (!traces.branch.trace.isEmpty && row >= traces.main.step && row < traces.main.step + traces.branch.trace.size) {
	      return traces.branch.trace(row - traces.main.step)
	    }
	  }
	  ""
	}
      })
		
      traceTable.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)

      traceTable.addMouseListener(new MouseListener {
	override def mouseClicked(e : MouseEvent) = {
	  val column = traceTable.getSelectedColumn
	  val row = traceTable.getSelectedRow

	  if (column == 0) {
	    if (row < traces.main.trace.size) {

	      var work = true
	      
	      while (traces.branch.step > 0 && work)
  		work = quietStepBack

	      while (traces.main.step > row && work)
		work = quietStepBack

	      while (traces.main.step < row && work)
		work = quietStep
	    }
	  } else {
	    if (traces.branch.trace != null && row >= traces.main.step && row < traces.main.step + traces.branch.trace.size) {
	      var work = true

	      while (traces.main.step + traces.branch.step > row && work)
		work = quietStepBack

	      while (traces.main.step + traces.branch.step < row && work)
		work = quietStep
	    }
	  }
	}
	override def mouseEntered(arg0 : MouseEvent) = {}
	override def mouseExited(arg0 : MouseEvent) = {}
	override def mousePressed(arg0 : MouseEvent) = {}
	override def mouseReleased(arg0 : MouseEvent) = {}
      })

      traceTable.setDefaultRenderer(classOf[Object], new TableCellRenderer {
	def isActive(row : Int, column : Int) : Boolean = {
	  if (column == 0) {
	    if (traces.branch.trace.isEmpty)
	      return row == traces.main.step
	  } else {
	    if (row >= traces.main.step && row < traces.main.step + traces.branch.trace.size) {
	      return (row - traces.main.step) == traces.branch.step
	    }
	  }
	  
	  return false
	}

	override def getTableCellRendererComponent(table : JTable, value : Object, isSelected : Boolean, hasFocus : Boolean, row : Int, column : Int) : Component = value match {
	  case str : String => cellRenderer.apply(str, isActive(row, column))
	  case _ => null
	}
      })
      traceTable
    }

    def setTraces(traces : TracePair) = {
      this.traces = traces
    }


    def getNextEvent : String = {
      val TraceStep(t, i) = getCurrentTrace
      if(t.size == i)
	null
      else
	t(i)
    }
	
    def asSimControl : SimControl[Swing, String] = {
      new SimControl[Swing, String] {
	override def unfire = unsafeToSwing(SimulationTraceTable.this.quietStepBack)
	override def getNextEvent : Swing[String] = unsafeToSwing(SimulationTraceTable.this.getNextEvent)
	override def fire(event : String) = unsafeToSwing(userRequestedFire(event))
	override def canFire(event : String) : Swing[Boolean] = model.canFire(event)
      }
    }
}
